File: GoL/Model/Cell.m
//
//  Cell.m
//  GoL
//
//  Created by Georg Meyer on 27.04.14.
//
//

#import "Cell.h"
#import "CellState.h"
#import "CellNeighborhood.h"

@interface Cell ()

@property (strong, nonatomic) CellState *state;
@property (strong, nonatomic) CellNeighborhood *neighborhood;

@end

@implementation Cell

- (BOOL)isAlive {
    return [self.state isAlive];
}

- (NSString *)description {
    return self.isAlive ? @"o" : @"_";
}

- (void)addNeighborCell:(Cell *)cell {
    [self.neighborhood addCell:cell];
}

- (void)storeNextState {
    BOOL currentState = self.state.isAlive;
    BOOL nextState = [self.neighborhood getNextState:currentState];
    [self.state storeNextState:nextState];
}

- (void)applyNextState {
    [self.state applyNextState];
}

- (Cell *)initWithCellState:(id)cellState {
    self = [self init];
    if (self) {
        self.state = cellState;
    }
    return self;
}

- (id)init {
    self = [super init];
    if (self) {
        self.neighborhood = [[CellNeighborhood alloc] init];
    }
    return self;
}

+ (Cell *)createAlive {
    return [[self alloc] initWithCellState:[CellState createAlive]];
}

+ (Cell *)createDead {
    return [[self alloc] initWithCellState:[CellState createDead]];
}

@end

File: GoL/Model/Cell.swift
//
//  Cell.swift
//  GoL
//
//  Created by Georg Meyer on 04.06.14.
//
//

import Foundation

extension Model {
    class Cell {
        
        var cellState : CellState
        var cellNeighborhood : CellNeighborhood
        var description : String {
            return isAlive() ? "o" : "_"
        }
        
        
        init(cellState : CellState){
            self.cellState = cellState
            self.cellNeighborhood = CellNeighborhood()
        }
        
        func isAlive() -> Bool {
            return cellState.isAlive()
        }
        
        func die() {
            cellState.die()
        }
        
        func revive() {
            cellState.revive()
        }
        
        func addNeighborCell(cell:Cell){
            cellNeighborhood.addCell(cell)
        }
        
        func storeNextState() {
            let currentState = cellState.isAlive()
            let nextState = cellNeighborhood.getNextState(currentState)
            cellState.storeNextState(nextState)
        }
        
        func applyNextState() {
            cellState.applyNextState()
        }
        
        class func createAlive() -> Cell {
            return Cell(cellState: CellState.createAlive())
        }
        
        class func createDead() -> Cell {
            return Cell(cellState: CellState.createDead())
        }
    }
}

